torch/nn/modules/module.py

```diff
@@ -927,6 +952,7 @@ def __getattr__(self, name: str) -> Union[Tensor, 'Module']:
         if '_parameters' in self.__dict__:
             _parameters = self.__dict__['_parameters']
             if name in _parameters:
+                self.__dict__[name] = _parameters[name]
```

Comment:
Is this super unsafe? If people do something with `_parameters`, __getattribute__ will still happily return stale cached value? 
Also, there's an issue to (possible) replace special `__getattr__` with `__getattribute__`, and this caching won't work then #50292. 
I'm not saying we shouldn't do it, but changes to nn.Module are to be made very carefully.

Reply:
I'm also wary of this kind of thing. In a previous life I implemented a similar kind of attribute caching scheme for Keras and  it was quite hard to maintain and ensure correctness. Fundamentally I think this is reflecting the fact that `__getattr__` is overridden while `__getattribute__` is not.

```
class MyClass:
    pass

x = MyClass()
x.a = 1
%timeit x.a

class MyClass:
    def __getattribute__(self, name):
        return super().__getattribute__(name)

x = MyClass()
x.a = 1
%timeit x.a
```

```
10000000 loops, best of 3: 42.3 ns per loop
1000000 loops, best of 3: 372 ns per loop
```

I think we should look seriously at writing a base class in C to handle some of the low level fast path bookkeeping. It has downsides, of course, but so do intrusive caching schemes.

Reply:
Very possibly - I should have made that more clear :P I imagine that a "correct" solution might need to do something similar to the `CachedOrderedDict` above, and clear the `__dict__` of parameter values whenever `_parameters` gets updated.

Reply:
@robieta I don't know if the current solution is right, but considering the performance gains to be had, the downsides might be worth it.

Reply:
And if people move the module with `.to()` for example, then accessing attributes will not pick the updated Parameter.

Reply:
> @robieta I don't know if the current solution is right, but considering the performance gains to be had, the downsides might be worth it.
> 
> The main point of this PR was just to show what needs to be done to reduce the `nn.Module` overhead.

I absolutely agree that the performance wins are compelling. To my mind it's not a question of "should we harvest them?", but simply how best to.

Reply:
Also what will happen when you do `model.weight = foo`? Currently, the `__setattr__` only sets the one in `_parameter` and so this caching might get stale? And if only the one in `__dict__` is now updated, then the one in `_parameter` will get stale.